File: evals/refs/bf.py
import sys
def run(code, inp):
    pairs = {}
    stack = []
    for i, ch in enumerate(code):
        if ch == '[': stack.append(i)
        elif ch == ']':
            j = stack.pop(); pairs[i] = j; pairs[j] = i
    tape = [0] * 30000
    ptr = 0; ip = 0; ipos = 0
    out = []
    while ip < len(code):
        c = code[ip]
        if c == '>': ptr += 1
        elif c == '<': ptr -= 1
        elif c == '+': tape[ptr] = (tape[ptr] + 1) & 0xFF
        elif c == '-': tape[ptr] = (tape[ptr] - 1) & 0xFF
        elif c == '.': out.append(chr(tape[ptr]))
        elif c == ',':
            tape[ptr] = ord(inp[ipos]) if ipos < len(inp) else 0
            ipos += 1
        elif c == '[' and tape[ptr] == 0: ip = pairs[ip]
        elif c == ']' and tape[ptr] != 0: ip = pairs[ip]
        ip += 1
    return ''.join(out)

lines = sys.stdin.read().split('\n')
T = int(lines[0])
out_lines = []
for k in range(T):
    code = lines[1 + 2*k]
    inp = lines[2 + 2*k]
    out_lines.append(run(code, inp))
sys.stdout.write('\n'.join(out_lines) + '\n')

File: evals/refs/bf.rs
use std::io::{self, BufRead, Write, BufWriter};

fn run(code: &str, inp: &str) -> String {
    let cb = code.as_bytes();
    let n = cb.len();
    let mut pairs = vec![0usize; n];
    let mut stack: Vec<usize> = Vec::new();
    for i in 0..n {
        if cb[i] == b'[' { stack.push(i); }
        else if cb[i] == b']' {
            let j = stack.pop().unwrap();
            pairs[i] = j; pairs[j] = i;
        }
    }
    let mut tape = vec![0u8; 30000];
    let inp_bytes = inp.as_bytes();
    let mut ptr = 0usize; let mut ip = 0usize; let mut ipos = 0usize;
    let mut out = String::new();
    while ip < n {
        match cb[ip] {
            b'>' => ptr += 1,
            b'<' => ptr -= 1,
            b'+' => tape[ptr] = tape[ptr].wrapping_add(1),
            b'-' => tape[ptr] = tape[ptr].wrapping_sub(1),
            b'.' => out.push(tape[ptr] as char),
            b',' => {
                tape[ptr] = if ipos < inp_bytes.len() { inp_bytes[ipos] } else { 0 };
                ipos += 1;
            }
            b'[' => { if tape[ptr] == 0 { ip = pairs[ip]; } }
            b']' => { if tape[ptr] != 0 { ip = pairs[ip]; } }
            _ => {}
        }
        ip += 1;
    }
    out
}

fn main() {
    let stdin = io::stdin();
    let stdout = io::stdout();
    let mut out = BufWriter::new(stdout.lock());
    let mut lines = stdin.lock().lines();
    let t: usize = lines.next().unwrap().unwrap().trim().parse().unwrap();
    for _ in 0..t {
        let code = lines.next().unwrap().unwrap();
        let inp = lines.next().unwrap().unwrap();
        writeln!(out, "{}", run(&code, &inp)).unwrap();
    }
}
